### PR TITLE
Debugging for 1.9

### PR DESCRIPTION
**Pull request by joel-hamill**
## What type of changes does your doc introduce?
- [ ] Bug fix
- [x] New feature 

## What is the urgency level of this change?
- [x] Blocker
- [x] High
- [ ] Medium

## I have completed these items:
- [x] I have tested all commands and code snippets.
- [x] I have built locally and tested for broken links and formatting.
- [ ] I have made this change for all affected versions (e.g. 1.7, 1.8, and 1.9).
- [ ] My doc follows the style guide: https://github.com/dcos/dcos-docs#contributing.

Ping `@emanic`, `@sascala`, or `@joel-hamill` for reviewing pull request changes.

### CODE REVIEW

File: 1.9/administration/debugging/quickstart.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    ```
+    
+    The output should look similar to this: 
+    
+    ```bash
+    NAME                                HOST       USER  STATE  ID                                                                       
+    20161209183121nz2F5.klueska-test    10.0.2.53  root    R    task-exec-test_<task_id>
+    ```
+
+1.  Launch a TTY process inside of the pod container with the task ID (<task_id>) specified. This will launch an interactive Bash session.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: There is no pod in this example, just a task

File: 1.9/administration/debugging/quickstart.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+Use this guide to get started with the `dcos exec` debugging command.
+
+**Prerequisite:**
+
+- A container launched by using the [DC/OS Universal container runtime](/docs/1.9/usage/containerizers/).
+
+
+# Launch a long running interactive Bash session
+
+In this example, a long running job app is launched and then the `dcos exec` command is used to launch an interactive Bash session inside of the task container on the node.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I wouldn't call it a `job app` I would just call it a `job` and maybe link to any metronome docs you have.

Additionally, I would probably reword this whole sentence as:
```
In this example, a long running job is launched via the `dcos job run` command and we use `dcos task exec` to launch an interactive Bash shell inside that job's container.
```

File: 1.9/administration/debugging/quickstart.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+1.  Deploy and run a job app with the DC/OS CLI:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Again, I wouldn't call it a `job app` that sounds confusing.

File: 1.9/administration/debugging/quickstart.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    1.  Create the following app definition and save as `⁠⁠⁠⁠test-job.json`. This specifies a sleep job that runs for `10000000` seconds.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: This is a `job` definition, not an `app` defintition.

File: 1.9/administration/debugging/quickstart.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            },
+            "restart": {
+              "policy": "NEVER"
+            },
+            "volumes": []
+          }
+        }
+        ```
+    
+    1.  Deploy the job app with this CLI command:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Just `job` not `job app`

File: 1.9/administration/debugging/quickstart.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    
+        ```bash
+        $ dcos job add test-job.json
+        ```
+        
+    1.  Verify that the app has been successfully deployed:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Verify the `job` has been successfully deployed

File: 1.9/administration/debugging/quickstart.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    ```bash
+    $ dcos task
+    ```
+    
+    The output should look similar to this: 
+    
+    ```bash
+    NAME                                HOST       USER  STATE  ID                                                                       
+    20161209183121nz2F5.klueska-test    10.0.2.53  root    R    <task_id>
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I would replace the `klueska-test` part of the name with `task-exec-test`

File: 1.9/administration/debugging/quickstart.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        task-exec-test               Unscheduled         None
+        ```
+
+    1.  Run the job:
+    
+        ```bash
+        $ dcos job run task-exec-test
+        ```
+
+1.  Get the task ID of the job with this CLI command:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Why is everything numbered as "1."?

File: 1.9/administration/debugging/quickstart.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    ```
+    
+    The output should look similar to this: 
+    
+    ```bash
+    NAME                                HOST       USER  STATE  ID                                                                       
+    20161209183121nz2F5.klueska-test    10.0.2.53  root    R    <task_id>
+    ```
+
+1.  Launch a TTY process inside of the container with the task ID (`<task_id>`) specified. This will launch an interactive Bash session.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I would reword this as:
```
Launch a process inside of the container with Task ID (`<task_id>`) and attach a TTY to it. This will launch an interactive Bash session.
```

File: 1.9/administration/debugging/quickstart.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    $ dcos task exec --interactive --tty <task_id> bash
+    ```
+    
+    You should now be inside the container running an interactive Bash session.
+    
+    ```bash
+    root@ip-10-0-2-53 / #
+    ```
+    
+    **Tip:** You can use shorthand abbreviations `-i` for `--interactive` or `-t` for `--tty`. Also, only the beginning unique characters of the `<task_id>` are required. For example, if your task is named `exec-test_20161214195` and there are no other task IDs that begin with the letter `e`, this is valid command syntax: `dcos task exec -i -t e bash`. For more information, see the CLI command [reference](/docs/1.9/usage/cli/command-reference/).
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I would say "if your task ID is `exec-test_20161214195`" instead of "if your task is named `exec-test_20161214195`"

File: 1.9/administration/debugging/quickstart.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    
+1.  Run a command from the interactive Bash session from inside the cluster. For example, the `ls` command:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I wouldn't both saying "from inside the cluster".

Just "Run a command in the interactive Bash session"

File: 1.9/administration/debugging/quickstart.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+    ```bash
+    root@ip-10-0-1-104 / # ls
+    bin   dev  home  lib64	     media  opt   root	sbin  sys  usr
+    boot  etc  lib	 lost+found  mnt    proc  run	srv   tmp  var
+    ```
+
+# Pipe output from a command running inside a container 
+
+You can run commands inside a container by using the `dcos exec` command. In this example, a long running Marathon app is launched and then the `dcos task exec` command is used to get the hostname of the node running your app.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: It's `dcos task exec` not `dcos exec`

Also there is a mismatch in voice here. You say "is launched" followed by "your app". I would probably change the latter to simply "the app".

File: 1.9/administration/debugging/quickstart.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    $ dcos task exec <task_id> hostname
+    ```
+    
+    The output should look similar to this:
+    
+    ```bash
+    ip-10-0-1-105.us-west-2.compute.internal
+    ```
+
+For more information about the `dcos task exec` command, see the CLI command [reference](/docs/1.9/usage/cli/command-reference/).
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I would probably add a reference directly to this in `1.9/administration/debugging/index.md` as well

File: 1.9/administration/debugging/quickstart.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+# Run an interactive command on a remote machine
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: It's not about running an interactive command on a remote machine. It's about running an interactive command inside your task's container (regardless of the machine it's on)

File: 1.9/administration/debugging/quickstart.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+You can run interactive commands on machines in your cluster by using the `dcos exec command`. In this example, the `dcos exec` command is used to copy a simple script from your local machine to the task container on the node. The script is then administered locally by using the `dcos exec` command.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Its the `dcos task exec` command, not just `dcos exec`

File: 1.9/administration/debugging/quickstart.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+            "placement": {
+              "constraints": []
+            },
+            "restart": {
+              "policy": "NEVER"
+            },
+            "volumes": []
+          }
+        }
+        ```
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Do we need such a complicated config. Can we make it simpler:
```
{
    "id": "task-exec-test",
    "container": {
        "type": "MESOS",
        "docker": {
            "image": "ubuntu:14.04"
        }
    },
    "cmd": "sleep 999999999",
    "cpus": 0.1,
    "instances": 1
}
```

File: 1.9/administration/debugging/quickstart.md [hunk 17]
```diff
@@ -81,19 +81,145 @@ In this example, a long running job app is launched and then a TTY process is la
...
+    
+    ```bash
+    NAME        HOST        USER  STATE  ID                                               
+    my-app  10.0.1.106  root    R    my-app.<task-ID>
+    ```
+
+1.  Run this command to copy a file to your cluster node. In this example a Marathon app definition is added. 
+ 
+    ```bash
+    $ cat my-app-cli.json | dcos task exec --interactive my-app-cli.a15eb2ae-c232-11e6-a451-aa711cbcaa78 cat my-app-cli.json
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: This command doesn't look right. I think it needs to be:
```
cat my-app-cli.json | dcos task exec --interactive my-app-cli.a15eb2ae-c232-11e6-a451-aa711cbcaa78 bash -c "cat > my-app-cli.json"
```

File: 1.9/administration/debugging/quickstart.md [hunk 18]
```diff
@@ -81,19 +81,145 @@ In this example, a long running job app is launched and then a TTY process is la
...
+    ```
 
- For more information about the `dcos task exec` command, see the CLI command [reference](/docs/1.9/usage/cli/command-reference/).
\ No newline at end of file
+an interactive, non-tty-based command to feed data into the command running on the remote machine (e.g. you can use `cat some-file | dcos task exec —interactive <task_id> cat - some-file` to copy a file into the remote container).
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I think something may have gotten lost here?

File: 1.9/administration/debugging/quickstart.md [hunk 19]
```diff
@@ -0,0 +1,241 @@
+---
+post_title: Quick Start
+menu_order: 3.3
+---
+
+Use this guide to get started with the `dcos exec` debugging command.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: This should be `dcos task exec`

File: 1.9/administration/debugging/quickstart.md (hunk 8 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: The markdown processor figures out the actual order during the build - that way we don't manually keep the numbers in sync.

File: 1.9/administration/debugging/quickstart.md (hunk 7 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Turning down fame huh?

File: 1.9/administration/debugging/quickstart.md (hunk 13 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Yep, this is already included further down in this PR.

File: 1.9/administration/debugging/quickstart.md (hunk 15 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Fixed globally.

File: 1.9/administration/debugging/quickstart.md (hunk 7 above)
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Haha. I'm all for calling it `klueska-test` throughout, so long as it's consistent everywhere :)

File: 1.9/administration/debugging/quickstart.md (hunk 16 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Definitely

File: 1.9/administration/debugging/quickstart.md (hunk 17 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: This is an old version of the doc.

File: 1.9/administration/debugging/quickstart.md (hunk 18 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: this is an old version of the doc

File: 1.9/administration/debugging/quickstart.md [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+    ```bash
+    $ cat hello-world.sh | dcos task exec -i <task_id> bash -c "cat > hello-world.sh"
+    ```
+    
+1.  Give the file executable permissions:
+
+    ```bash
+    $ dcos task exec <task_id> chmod a+x hello-world.sh
+    ```
+1. Run the script inside of the container:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Newline above this?

File: 1.9/administration/debugging/quickstart.md [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+
+    ```bash
+    {
+       "id": "/my-app",
+       "cmd": "sleep 100000000",
+       "cpus": 1
+    }
+    ```
+
+1.  Deploy the service on DC/OS:
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Service or app? You say app everywhere else.

File: 1.9/administration/debugging/quickstart.md [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+
+    The output should look similar to this:
+    
+    ```bash
+    Hello World
+    ```
+
+# Launch a long running interactive Bash session
+
+In this example, a long running [job](/docs/1.9/usage/jobs/) is launched by using the `dcos job run` command, and the `dcos task exec` command is used to launch an interactive Bash shell inside the container of that container.
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: `container of that container` should be `container of that job`

File: 1.9/administration/debugging/quickstart.md [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+    1.  Create the following job definition and save as `⁠⁠⁠⁠test-job.json`. This specifies a sleep job that runs for `10000000` seconds.
+    
+        ```bash
+        {
+            "id": "task-exec-test",
+            "cmd": "sleep 999999999",
+            "cpus": 0.1,
+            "instances": 1
+        }
+        ```
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: I would probably change this to have an id of `my-job` for consistency with the app definition in the previous section. Also, I don't think you can specify `instances` in job definitions (only in app definitions). Did you try this out yourself?

File: 1.9/administration/debugging/quickstart.md [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+        ```bash
+        $ dcos job add test-job.json
+        ```
+        
+    1.  Verify that the job has been successfully deployed:
+    
+        ```bash
+        $ dcos job list
+        ID              DESCRIPTION  STATUS   LAST SUCCESFUL RUN        
+        task-exec-test               Unscheduled         None
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: If we change the id to `my-job` above, make sure and propagate that throughout.

File: 1.9/administration/debugging/quickstart.md [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+1.  Create a Marathon app definition and name it `my-app.json` with the following contents:
+
+    ```bash
+    {
+       "id": "/my-app",
+       "cmd": "sleep 100000000",
+       "cpus": 1,
+       "instances": 1
+     }
+     ```
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Do we maybe want to rename this app to something else? If someone runs through these instructions from top to bottom, this app-id will conflict with the one already deployed by the first set of instructions and they won't be able to launch the app.

Also, I think this should probably be the first example in this document. I think it flows better from least complicated (stream output only) to most complicated (full interactive bash session).

Finally, This app definition (other than the `id`) should probably match the other one for consistency.

File: 1.9/administration/debugging/quickstart.md (hunk 21 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Changed to app

File: 1.9/administration/debugging/quickstart.md (hunk 23 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: I have fixed. I inadvertently severely messed up this job example. I have fixed.

File: 1.9/administration/debugging/quickstart.md [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+    
+    The output should look similar to this:
+    
+    ```bash
+    ip-10-0-1-105.us-west-2.compute.internal
+    ```
+
+For more information about the `dcos task exec` command, see the CLI command [reference](/docs/1.9/usage/cli/command-reference/).
+
+# Run an interactive command inside task container
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Change "inside task container" to "inside a task's container"

File: 1.9/administration/debugging/quickstart.md [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+You can run interactive commands on machines in your cluster by using the `dcos task exec` command. In this example, the `dcos task exec` command is used to copy a simple script from your local machine to the task container on the node. The script is then administered locally by using the `dcos task exec` command.
+
+1.  Create a Marathon app definition and name it `my-interactive-app.json` with the following contents:
+
+    ```bash
+    {
+       "id": "/my-interactive-app",
+       "cmd": "sleep 100000000",
+       "cpus": 1
+    }
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: The app definition for "/my-app" has `"instances": 1` set in it. I'd either add this here or remove it from above (the default is 1 if you don't specify anything). My preference is to remove it from above to make these definitions as simple as possible.

File: 1.9/administration/debugging/quickstart.md [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+    ```
+    
+    You should now be inside the container running an interactive Bash session.
+    
+    ```bash
+    root@ip-10-0-2-53 / #
+    ```
+    
+    **Tip:** You can use shorthand abbreviations `-i` for `--interactive` or `-t` for `--tty`. Also, only the beginning unique characters of the `<task_id>` are required. For example, if your task ID is `exec-test_20161214195` and there are no other task IDs that begin with the letter `e`, this is valid command syntax: `dcos task exec -i -t e bash`. For more information, see the CLI command [reference](/docs/1.9/usage/cli/command-reference/).
+
```
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Since this is the last thing in the section now, maybe move this tip to the very bottom so as not to break the flow?

File: 1.9/administration/debugging/quickstart.md (hunk 16 above)
Comment (klueska) [inline 1.9/administration/debugging/quickstart.md]: Did something accidentally get reverted here? I think the job definition should just be:
```
{
  "id": "my-job",
  "run": {
    "cpus": 0.01,
    "cmd": "sleep 99999"
  }
}

This launches something equivalent to what we specify in the marathon app definitions above.
```

Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: I used what worked (that long form). i don't have time to test other permutations at this point.

File: 1.9/administration/debugging/quickstart.md (hunk 27 above)
Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Added

File: 1.9/administration/debugging/quickstart.md [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+    ```bash
+    $ dcos task
+    ```
+    
+    The output should look similar to this: 
+    
+    ```bash
+    NAME        HOST        USER  STATE  ID                                               
+    my-app  10.0.1.106  root    R    <task_id>
```
Comment (adam-mesos) [inline 1.9/administration/debugging/quickstart.md]: This table does not look properly aligned.

File: 1.9/administration/debugging/quickstart.md [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+
+    ```bash
+    $ dcos task
+    ```
+    
+    The output should look similar to this: 
+    
+    ```bash
+    NAME        HOST        USER  STATE  ID                                               
+    my-interactive-app  10.0.1.106  root    R    <task_id>
```
Comment (adam-mesos) [inline 1.9/administration/debugging/quickstart.md]: Please align

File: 1.9/administration/debugging/quickstart.md [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+        ```bash
+        $ dcos job add my-job.json
+        ```
+        
+    1.  Verify that the job has been successfully deployed:
+    
+        ```bash
+        $ dcos job list
+        ID              DESCRIPTION  STATUS   LAST SUCCESFUL RUN        
+        my-job               Unscheduled         None
```
Comment (adam-mesos) [inline 1.9/administration/debugging/quickstart.md]: Please align

Comment (joel-hamill) [inline 1.9/administration/debugging/quickstart.md]: Fixed https://github.com/dcos/dcos-docs/pull/806
